### DISCUSSION TITLE
两个都是分类变量如何做ROC？

### DISCUSSION THREAD

**Discussion by zhaoliang0302** [ShixiangWang/self-study#99] [Q&A] [answered]
翔哥好，

有一个统计的问题，做roc时常规是连续变量和分类变量之间做，但是我现在想看基因甲基化与否和治疗反应（responder、non-responder）之间做roc，我搜索了好像这样做不科学。但是我目前有一个模型（连续变量）来预测治疗反应，做了roc，想知道怎么做才能在模型和基因甲基化之间的预测效能方面作比较？谢谢

**Comment by ShixiangWang** [comment, marked as answer]
@zhaoliang0302 有2种方式：

1. 对甲基化数据建模后的概率值预测治疗反应，概率值是连续的
2. 不使用ROC，而是对不同模型确定一个二分类的阈值，对比不同模型最佳阈值下的准确度，精度，召回等等。

**Reply by zhaoliang0302** [reply]
第一种方式可能不适用，甲基化只是二分类（甲基化、非甲基化）资料，没有那种beta值类似的连续数据；
第二种方式没明白意思，有推荐的教程可以学习下吗？
谢谢

**Reply by ShixiangWang** [reply]
第一种跟数据类型没有关系，大部分模型，比如逻辑回归，输入类型都可以，输出是2分类，建模后预测数据是可以得到概率值，这个概率值可以作为一个连续score产生ROC之类的曲线。

第二种也跟第一种有关，但利用的不是概率值，而是输出的分类，通常阈值是概率为0.5，即>0.5分到正类。后面可以生成一个混淆矩阵，就可以计算各种性能评估。

**Reply by zhaoliang0302** [reply]
我的数据样本只有20多个，构建逻辑回归模型再分测试集验证集可能太少了，我可能选第二种方法。因为甲基化与否的信息有了，只需要把模型连续变量分成二分类，你的意思是按50%比例对半分？

**Reply by ShixiangWang** [reply]
那你这种不就是一个2x2表格？比如 A 甲基化 YES/NO - 治疗反应 YES/NO

**Reply by ShixiangWang** [reply]
一般这种就用fisher.test算个关联就可以了，比如说A基因甲基化富集在反应组。

**Reply by zhaoliang0302** [reply]
主要是想把甲基化和自己的模型比较他们预测治疗反应的效能，甲基化这个是经典的影响治疗的因素，我想说明我的模型效果比甲基化的好

**Reply by ShixiangWang** [reply]
那就建议用第一种，你只需要建模做个回归，不需要预测。甚至更简单的，你把甲基化分为高低编码为1和0，然后做ROC曲线也是可以的。区别在于你做个逻辑回归得到的概率值生成曲线更圆滑。

**Reply by zhaoliang0302** [reply]
我尝试了一下：
```r
head(data)
# A tibble: 6 x 2
#  methylation     respond
#   <chr>          <dbl>
# methylated         0
# methylated         0
# methylated         1
# unmethylated       0
# unmethylated       1
# methylated         0

model_logit <- glm(respond ~ methylation,
                    data = data,
                    family = 'binomial'
 )
prop <- predict(model_logit)

length(unique(predict(model_logit)))
# 2
```
得到的虽然是数值，但是只有两种数值，ROC曲线也是尖锐的算出来的，AUC和直接把甲基化分为高低编码为1和0是一样的
![image](https://user-images.githubusercontent.com/42333702/177433564-761d7693-0004-4741-add7-3fe6fadaa6aa.png)

**Reply by ShixiangWang** [reply]
是的，因为R内置进行了转换。你如果用它作为一个对比是没有问题的，发文章别人问你你也可以直接解释。

**Reply by zhaoliang0302** [reply]
非常感谢，为你开放交流的精神点赞👍

**Reply by ShixiangWang** [reply]
不客气
